294-documenting-assignment-process

andreievg opened the pull request:
Fixes: #294 

Added new folder to drive, IRIMIS/Internal `Documentation Diagrams/Web App/`, will be relocating documentation diagrams there (as documentation is PRed)

Also added a new diagram `Assignment UI Flow`

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+# Assignment Process
+
+The logic of assigning a review spans between both web app and server, front end is only concerned with queries and mutations to review_assignment and review_question_assignment tables, whereas back end does most of the grunt work (through actions and triggers) and generates records in those tables based on user permissions and states of application and it's review.
```
```suggestion
The logic of assigning a review spans between both web app and server — front end is only concerned with queries and mutations to review_assignment and review_question_assignment tables, whereas back end does most of the grunt work (through actions and triggers) and generates records in those tables based on user permissions and states of application and its review.
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+First of all, some bulletpoint (and rules) to help clarify the diagram
+
+- One review assignment should exists per reviewer/application/stage/level
```
```suggestion
- One review assignment should exist per reviewer/application/stage/level
- ```

andreievg replied:
f74632106a6a1af38d301b000f3fb8027bf7178e
did a manual commit, i saw an extra line in the suggested change

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+- Review assignment can have the following statuses
+  - `Available` -> Can be assigned by assigner
+  - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
+  - `Assigned` -> At least on review_question_assignment was linked to review_assignment
```
```suggestion
  - `Assigned` -> At least one review_question_assignment was linked to review_assignment
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+  - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
+  - `Assigned` -> At least on review_question_assignment was linked to review_assignment
+  - `Available for self-assignment` -> Can be self assigned
+- Review assignments can be restricted by section (template_section_restriction, which is an array of section codes)
```
Maybe clarify here that "restrictions" means the sections that you *can* do, but if there's not restrictions specified, they can do *all* sections.

andreievg replied:
5d80e7de004763c37418086f20a9e44f0ee99f87

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+  - `Assigned` -> At least on review_question_assignment was linked to review_assignment
+  - `Available for self-assignment` -> Can be self assigned
+- Review assignments can be restricted by section (template_section_restriction, which is an array of section codes)
+- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the join table
```
```suggestion
- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the `review_assignment_assigner_join` table
```
Maybe?

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+  - `Available for self-assignment` -> Can be self assigned
+- Review assignments can be restricted by section (template_section_restriction, which is an array of section codes)
+- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the join table
+- `Assigned` status is used to indicate to the reviewer that they can start the review (if review hasn't been started already)
```
```suggestion
- `Assigned` status indicates that the reviewer is able to create a review (or one already exists)
```

nmadruga commented:
Yey 🎉  - Will review this very soon - still trying to get all PRs from previous week in

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+- Review assignments can be restricted by section (template_section_restriction, which is an array of section codes)
+- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the join table
+- `Assigned` status is used to indicate to the reviewer that they can start the review (if review hasn't been started already)
+- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignmetn is different to application stage
```
```suggestion
- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignmetn is different to application stage
+- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)
+
+## Difference between self assignemnt and assignment
```
```suggestion
## Difference between self-assignment and assignment
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+## Difference between self assignemnt and assignment
+
+The idea behind self assignment originated with requirement for users (who are not themself assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions
```
```suggestion
The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions.
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+## The Act of Assigning - Web App
+
+From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, changes status of review_assignment to `Assigned`. For self assignemnt trigger is set to `ON_REVIEW_SELF_ASSIGN`.
```
```suggestion
From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, and changes status of review_assignment to `Assigned`. For self assignment, trigger is set to `ON_REVIEW_SELF_ASSIGN`.
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+To reduce front end logic, server is responsible for generating review_assignment records, based on user permissions. This is done by an action `generateReviewAssignments` which is configured to fire when application or review changes status.
+
+Basically, when application is submitted, we want to look at all of the permissions linked to application template of type Assign, where template permission configurations match current level and stage, review_assignment records are generated for users who are linked to that template permission (through permission name), and any extra configurations (template section restrictions, canSelfAssign,  ...
```
I don't think this is quite correct -- we look for permissions of type "Review" to create the review_assignment records. We look at "Assign" permissions to create review_assignment_assigner_join records.

andreievg replied:
0cdebd150db218cf5ab00558507ae5617e29b0a3
oops, thanks for noting

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+## Diagram
+
+This diagram is somewhat complex, it describes some functionality that is not implemented (re-assignment), it can be helpful when visualising full assignment flow (and for analysing different scenariors of assignment)
```
```suggestion
This diagram is somewhat complex — it describes some functionality that is not implemented (re-assignment), but it can be helpful when visualising full assignment flow (and for analysing different scenarios of assignment)
```

CarlosNZ commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+Best described in a diagram (this diagram is abstracted to only show bits related to assignments)
+
+![Assignment UI Flow](images/Assignment-Flow-Detailed.png)
```
Links to the same image as the previous -- I think you probably meant "Assignment-UI-Flow.png"?

andreievg replied:
17356cda07839e3fb12cb9b10035e275b16d6836

CarlosNZ reviewed (changes requested):
Great, thanks for summarising this. Just a bunch of minor typos and suggestions.

andreievg commented:
@CarlosNZ thanks for you review, sorry for so many typos, I'll try to review my own doc PR before submitting

CarlosNZ reviewed (approved):
Sweet, thanks for the tidy-ups. :)

nmadruga commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+First of all, some bulletpoint (and rules) to help clarify the diagram
+
+- One review assignment should exist per reviewer/application/stage/level
+- Review assignment can have the following statuses
+  - `Available` -> Can be assigned by assigner
+  - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
+  - `Assigned` -> At least one review_question_assignment was linked to review_assignment
+  - `Available for self-assignment` -> Can be self assigned
+- Review assignments can be restricted by section/s (template_section_restriction, which is an array of section codes user can review, value can be blank to allow review of all sections)
+- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the `review_assignment_assigner_join` table
+- `Assigned` status indicates that the reviewer is able to create a review (or one already exists)
+- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
+- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)
```
Suggestion to use sub-headers to break-down get the subject of each one of this important points/rules...

```suggestion
### Assignment rules
- One `review_assignment` should exist per reviewer/application/stage/level
- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)

### Assignment status
Review assignment can have the following statuses
 - `Available` -> Can be assigned by assigner
 - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
 - `Assigned` -> indicates that the reviewer is able to create a review (or one already exists)
   - At least one `review_question_assignment` was linked to review_assignment
 - `Available for self-assignment` -> Can be self assigned

### Restrictions
Review assignments can be restricted by section/s
 - `template_section_restriction` is an array of section codes user can review
 - Default is `null` which allow review of all sections or all questions in the application
 
### Assigner settings
- Review assignments of status `Available` or `Assigned` can be assigned by an assigner
- Assigner is linked through the `review_assignment_assigner_join` table

```

nmadruga replied:
Have added sub-header and moved things up and down a little - see if that's helpful. Also I couldn't figure out from this description 2 things:

1. What is the conclusion of this sentence: "Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage."
2. "Assigner is linked through the `review_assignment_assigner_join` table" How does this work? I was trying to understand this, that's why I'm asking here also.

nmadruga reviewed (changes requested):
Just started reviewing this now.

nmadruga commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
+- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)
+
+## Difference between self-assignment and assignment
+
+The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions.
```
Following up what is suggested previously

```suggestion
#### Difference between self-assignment and assignment

The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. 

**Note**: Currently self assignment is not restricted by `template_section_restrictions`.
```

nmadruga commented on documentation/internal/Assignment-Process.md:
```diff
@@ -0,0 +1,57 @@
+
+The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions.
+
+## The Act of Assigning - Web App
+
+From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, and changes status of review_assignment to `Assigned`. For self assignment, trigger is set to `ON_REVIEW_SELF_ASSIGN`.
+
+Front end doesn't generate review_assignment records, this happens on the server.
+
+## Server Side
+
+To reduce front end logic, server is responsible for generating review_assignment records, based on user permissions. This is done by an action `generateReviewAssignments` which is configured to fire when application or review changes status.
+
+Basically, when application is submitted, we want to look at all of the permissions linked to application template of type Review, where template permission configurations match current level and stage, review_assignment records are generated for users who are linked to that template permission (through permission name), and any extra configurations (template section restrictions, canSelfAssign,  ...
+
+When reviews are submitted, we do similar thing but for level + 1. Typical scenario is a consolidator being informed that they can self assign a consolidation when lower level reviewer submits a review.
```
Moved things a lot on this suggestion. Basically deals first with what was in "Server side" with title "Pre-assignment process" and list what happens on the front end under the same title "Web app" but now with sub-titles of "Self-assignment" and "Assigning to others".

```suggestion

## Pre-assignment process

To reduce front-end logic, server is responsible for generating `review_assignment` records, based on user permissions. This is done by an action called `generateReviewAssignments` which is configured to fire when application or review changes status.
e.g.: Application is Submitted, Review is approved and moves to new Stage.

### Application submission
The server logic looks at all of the permissions linked to application template of type `review`, where template permission configurations match current `level` and `stage`. It creates `review_assignment` records for users who are linked to that template permission (through permission name), and any extra configurations (`template_section_restrictions`, `canSelfAssign`, etc..) are populated for new review_assignment record.

### Review submission
The server logic does similar thing as what is done when Application is submitted, but for level + 1. Typical scenario is a consolidator being informed that they can self assign a consolidation when lower level reviewer submits a review.

## Assigning process
Available for either a Reviewer that can do **self-assignment** or an Assigner with permission to **assign** applications of a certain template type on a `level` and `stage` (settings defined on the `template_permission` of each Template).

 To **assign** or **self-assignment** the Reviewer of sections of one application there are option available on the Applications list of on the Application home page.  

### Web App

#### Self-assignment
A GraphQL mutation is created the front-end when the user clicks on the **Self-assign** for one section.
What happens as a result of this mutation is: 
1. `review_assignment.trigger` is set to ON_REVIEW_SELF_ASSIGN.
2. Changes `review_assignment.status` to **Assigned**. 
3. Adds `review_question_assignment` (for each question of section assigned) to `review_assignment`
4. Other updates done on the server-side:
  - Update other `review_assignment.status` from **Available for self-assignment** to **Self-assigned by another**
  - Assign other sections to the same reviewer - to avoid misleading half-review.

#### Assigning to others
A GraphQL mutation  is created the front-end when the user selects one Reviewer on the drop-down list to **assign** for one section. What happens as a result of this mutation is: 
1. `review_assignment.trigger` is set to ON_REVIEW_ASSIGN.
2. Changes status of `review_assignment` to **Assigned**.
3. Adds `review_question_assignment` (for each question on the section assigned) to `review_assignment`
4. Other updates done on the server-side:
  - Update other `review_assignment.status` from **Available** to **Locked**
```

nmadruga reviewed (changes requested):
Ok. My review is finished. Sorry, have added quite long suggestions, I just feel like it's easier to refer to something when there as subtitles, and although the diagrams can be helpful - I have checked something on it while making this review, it's a bit complex to look to find a small thing.

nmadruga commented:
On last overall comment - I added the swap to status **Locked** when the review is assigned to another person, but that's actually a `review.status` and not one `review_assignment.status`, but I think it would be good to define this here.

nmadruga commented:
Hi @andreievg is this review good for another review? I wasn't sure, but saw you made changes 2 weeks ago...

nmadruga commented:
@CarlosNZ @andreievg I suggest we merge this in - even without latest requests and go improving as we go. Because it is way better to have **some** documentation available then none. And we will be using all of these (most related to next PRs) for consulting and updating.

andreievg commented:
> Hi @andreievg is this review good for another review? I wasn't sure, but saw you made changes 2 weeks ago...
 
 > @CarlosNZ @andreievg I suggest we merge this in - even without latest requests and go improving as we go. Because it is way better to have some documentation available then none. And we will be using all of these (most related to next PRs) for consulting and updating.
 
I was bumping this as low priority. I'll try to address this soon (today/tomorrow)

andreievg commented:
> On last overall comment - I added the swap to status Locked when the review is assigned to another person, but that's actually a review.status and not one review_assignment.status, but I think it would be good to define this here.

I dont' think this is happening.

I've added suggested changes and made some changes to those : )

c85058941db62bc2026812c2ddcb12668bc799a7

I think all of the comments are now addressed
